File: mobile/ios/Vault/ApiClient.swift
import Foundation

/// Minimal server client (URLSession). Access tokens live only in memory.
final class ApiClient {
    let baseUrl: String
    var accessToken: String?
    private let session = URLSession(configuration: .ephemeral)

    init(baseUrl: String) { self.baseUrl = baseUrl }

    func loginStart(username: String, credentialRequest: String) async throws -> [String: Any] {
        try await post("/auth/login/start", ["username": username, "credential_request": credentialRequest])
    }

    func loginFinish(flowId: String, finalization: String, deviceName: String, totp: String?) async throws -> [String: Any] {
        var body: [String: Any] = [
            "flow_id": flowId, "credential_finalization": finalization, "device_name": deviceName
        ]
        if let totp { body["totp_code"] = totp }
        return try await post("/auth/login/finish", body)
    }

    func accountCrypto() async throws -> [String: Any] { try await get("/account/crypto", auth: true) }
    func pull(cursor: Int) async throws -> [String: Any] { try await get("/sync?cursor=\(cursor)", auth: true) }
    func push(record: [String: Any], baseVersion: Int) async throws -> [String: Any] {
        try await post("/sync/push", ["record": record, "base_version": baseVersion], auth: true)
    }

    private func post(_ path: String, _ body: [String: Any], auth: Bool = false) async throws -> [String: Any] {
        var req = URLRequest(url: URL(string: "\(baseUrl)/api/v1\(path)")!)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if auth, let t = accessToken { req.setValue("Bearer \(t)", forHTTPHeaderField: "Authorization") }
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        return try await run(req)
    }

    private func get(_ path: String, auth: Bool = false) async throws -> [String: Any] {
        var req = URLRequest(url: URL(string: "\(baseUrl)/api/v1\(path)")!)
        if auth, let t = accessToken { req.setValue("Bearer \(t)", forHTTPHeaderField: "Authorization") }
        return try await run(req)
    }

    private func run(_ req: URLRequest) async throws -> [String: Any] {
        let (data, resp) = try await session.data(for: req)
        let obj = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]
        if let http = resp as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            throw VaultError.Message(message: obj["error"] as? String ?? "http \(http.statusCode)")
        }
        return obj
    }
}

File: mobile/android/app/src/main/java/au/com/rodoskosmos/vault/net/ApiClient.kt
package au.com.rodoskosmos.vault.net

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject

/** Minimal server client. Access tokens live only in memory. */
class ApiClient(var baseUrl: String) {
    private val http = OkHttpClient()
    private val json = "application/json".toMediaType()
    var accessToken: String? = null

    private fun post(path: String, body: JSONObject, auth: Boolean = false): JSONObject {
        val builder = Request.Builder()
            .url("$baseUrl/api/v1$path")
            .post(body.toString().toRequestBody(json))
        if (auth) accessToken?.let { builder.header("Authorization", "Bearer $it") }
        return exec(builder.build())
    }

    private fun get(path: String): JSONObject {
        val builder = Request.Builder().url("$baseUrl/api/v1$path")
        accessToken?.let { builder.header("Authorization", "Bearer $it") }
        return exec(builder.build())
    }

    private fun exec(request: Request): JSONObject {
        http.newCall(request).execute().use { resp ->
            val text = resp.body?.string().orEmpty()
            val obj = if (text.isBlank()) JSONObject() else JSONObject(text)
            if (!resp.isSuccessful) throw ApiException(resp.code, obj.optString("error", "error"), obj)
            return obj
        }
    }

    fun loginStart(username: String, credentialRequest: String): JSONObject =
        post("/auth/login/start", JSONObject().put("username", username).put("credential_request", credentialRequest))

    fun loginFinish(flowId: String, finalization: String, deviceName: String, totp: String?): JSONObject =
        post(
            "/auth/login/finish",
            JSONObject()
                .put("flow_id", flowId)
                .put("credential_finalization", finalization)
                .put("device_name", deviceName)
                .apply { if (totp != null) put("totp_code", totp) }
        )

    fun accountCrypto(): JSONObject = get("/account/crypto")

    fun pull(cursor: Long): JSONObject = get("/sync?cursor=$cursor")

    fun push(record: JSONObject, baseVersion: Long): JSONObject =
        post("/sync/push", JSONObject().put("record", record).put("base_version", baseVersion), auth = true)
}

class ApiException(val status: Int, val code: String, val body: JSONObject) :
    Exception("$status $code")
